src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -110,16 +112,17 @@ private synchronized void open() throws IOException {
     }
 
     @Override public BuildListener overallListener() throws IOException, InterruptedException {
-        return new StreamBuildListener(new IndexOutputStream(null), StandardCharsets.UTF_8);
+        return new BufferedBuildListener(new IndexOutputStream(null));
     }
 
     @Override public TaskListener nodeListener(FlowNode node) throws IOException, InterruptedException {
-        return new StreamTaskListener(new IndexOutputStream(node.getId()), StandardCharsets.UTF_8);
+        return new BufferedBuildListener(new IndexOutputStream(node.getId()));
     }
 
     private void checkId(String id) throws IOException {
         assert Thread.holdsLock(this);
         if (!Objects.equals(id, lastId)) {
+            bos.flush();
```

Comment:
Maybe I misunderstand something here (trying to review fast),  but won't it defeat the point if we flush here, since `checkId` is called with interleaved parallels?  Along with that, wouldn't it make sense to buffer the index output stream too?

Reply:
Specifically b/c the index gets a *lot* of tiny writes with parallel interleaved logging.

Reply:
> won't it defeat the point if we flush here, since `checkId` is called with interleaved parallels

Not so often, because

> the index gets a _lot_ of tiny writes with parallel interleaved logging

is no longer true to nearly the same extent, due to the remote buffering.

It seemed unsafe to leave content not yet flushed to `log` when `log-index` already referred to that transition: your index would then contain offsets greater than the length of the file.

> wouldn't it make sense to buffer the index output stream too

Should not be necessary as `log-index` is orders of magnitude smaller than `log`.

Reply:
> Should not be necessary as log-index is orders of magnitude smaller than log.

Note that this may not *necessarily* be true, to give one example: parallel shell steps that log relatively slowly but constantly will generate a lot of log index entries until they hit the max backoff (one of the better arguments for a somewhat higher max backoff and growth factor).  For example, something like: 

> randomDownloadTool downloading artifact_foo.jar
> .
> .
> .
> .
> .
> download completed

Would generate this result.

However the other argument is compelling enough in general.